# Titanic Prediction

In this section we are going to predict wether a person `survived` the titanic incident or not using **Logistic Regression** model with 6 features.

## Datasets

For the datasets, we are going to use datas available from this source

https://sololearn.com/uploads/files/titanic.csv

In [33]:
import pandas as pd

# Dataframe
df = pd.read_csv('https://sololearn.com/uploads/files/titanic.csv')
df.head()

Survived  Pclass     Sex   Age  Siblings/Spouses  Parents/Children     Fare
0         0       3    male  22.0                 1                 0   7.2500
1         1       1  female  38.0                 1                 0  71.2833
2         1       3  female  26.0                 0                 0   7.9250
3         1       1  female  35.0                 1                 0  53.1000
4         0       3    male  35.0                 0                 0   8.0500

## Formatting Data

Next, we are going to format the data above, so that it is easier to be used for training. Notice how the `Sex` column have a **custom** value. We need to format this feature column immediately to prevent errors in training later on.

In [34]:
df['Male'] = df['Sex'] == 'male'
df[['Survived', 'Pclass', 'Male', 'Age', 'Siblings/Spouses', 'Parents/Children', 'Fare']].head()

Survived  Pclass   Male   Age  Siblings/Spouses  Parents/Children     Fare
0         0       3   True  22.0                 1                 0   7.2500
1         1       1  False  38.0                 1                 0  71.2833
2         1       3  False  26.0                 0                 0   7.9250
3         1       1  False  35.0                 1                 0  53.1000
4         0       3   True  35.0                 0                 0   8.0500

We change the `Sex` feature column to `Male` and succesfully changed the data type of the datasets to **Boolean**. Next, let's convert it to an array and split the target column from the feature column as well.

In [35]:
X = df[['Pclass', 'Male', 'Age', 'Siblings/Spouses', 'Parents/Children', 'Fare']].values
y = df['Survived'].values
print(X)
print(y)

[[3 True 22.0 1 0 7.25]
 [1 False 38.0 1 0 71.2833]
 [3 False 26.0 0 0 7.925]
 ...
 [3 False 7.0 1 2 23.45]
 [1 True 26.0 0 0 30.0]
 [3 True 32.0 0 0 7.75]]
[0 1 1 1 0 0 0 0 1 1 1 1 0 0 0 1 0 1 0 1 0 1 1 1 0 1 0 0 1 0 0 1 1 0 0 0 1
 0 0 1 0 0 1 1 0 0 1 0 0 0 0 1 1 0 1 1 0 1 0 0 1 0 0 0 1 1 0 1 0 0 0 0 0 1
 0 0 0 1 1 0 1 1 0 1 1 0 0 1 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 1 1 0 1 0 0
 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 1 1 0 0 0 0 1 0 0 1 0 0 0 0 1 1 0 0 0 1 0 0
 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 1 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 1 0
 1 1 0 0 1 0 1 1 1 1 0 0 1 0 0 0 0 0 1 0 0 1 1 1 0 1 0 0 0 1 1 0 1 0 1 0 0
 0 1 0 1 0 0 0 1 0 0 1 0 0 0 1 0 0 0 1 0 0 0 0 0 1 1 0 0 0 0 0 0 1 1 1 1 1
 1 0 0 0 0 0 1 1 1 0 1 1 0 1 1 0 0 0 1 0 0 0 1 0 0 1 0 1 1 1 1 0 0 0 0 0 0
 1 1 1 1 0 1 0 1 1 1 0 1 1 1 0 0 0 1 1 0 1 1 0 0 1 1 0 1 0 1 1 1 1 0 0 0 1
 0 0 1 1 0 1 1 0 0 0 1 1 1 1 0 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 0 1 1 1 1 1 0
 0 0 0 1 1 0 0 0 1 1 0 1 0 0 0 1 0 1 1 1 0 1 1 0 0 0 0 1 1 0 0 0 0 0 0 1 0
 0 0 0 1 0 1 0 1 1

## Training

The data is already well prepared and ready for training. To do so, we can use the model provided from `scikit learn` to train our data.

In [36]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X, y)

# Print equation coefficient
print(model.coef_, model.intercept_)

[[-1.13641993 -2.64406541 -0.04237437 -0.38747099 -0.0962051   0.00297279]] [5.08847342]


## Verify

Now let's verify the model by trying to predict the first 5 data

In [37]:
print(model.predict(X[:5]))

[0 1 1 1 0]


In [38]:
print(y[:5])

[0 1 1 1 0]


As we can see, our model succesfully predicts the first 5 data. Next, we can try to find the accuracy score of our model.

In [39]:
print(model.score(X, y))

0.8049605411499436


This shows, that our model has 80% accuracy, which mean the predicted answer is 80% correct

# Breast Cancer Prediction

In this section we are going to predict wether a person have a `breast cancer` or not using **Logistic Regression** model with 10 attributes, with each attribute is split in to 3 features (mean, standard error, "worst" or "largest"). This in total gives the datasets 30 features to work with.

## Datasets

For the datasets, we are going to use datas available from `scikit` itself

In [40]:
from sklearn.datasets import load_breast_cancer

cancer_data = load_breast_cancer()
print(cancer_data.keys())

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


We can get more information about this dataset by taking a look at the description

In [41]:
print(cancer_data['DESCR'])

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

:Number of Instances: 569

:Number of Attributes: 30 numeric, predictive attributes and the class

:Attribute Information:
    - radius (mean of distances from center to points on the perimeter)
    - texture (standard deviation of gray-scale values)
    - perimeter
    - area
    - smoothness (local variation in radius lengths)
    - compactness (perimeter^2 / area - 1.0)
    - concavity (severity of concave portions of the contour)
    - concave points (number of concave portions of the contour)
    - symmetry
    - fractal dimension ("coastline approximation" - 1)

    The mean, standard error, and "worst" or largest (mean of the three
    worst/largest values) of these features were computed for each image,
    resulting in 30 features.  For instance, field 0 is Mean Radius, field
    10 is Radius SE, field 20 is Worst Radius.

    - 

After understanding the description of the data set, we know that there is a total of 30 features to work with. Another import information is, `Malignant` indicates that the breast is cancerous, while `Benign` indicates that the breast is **not** cancerous.

In [42]:
print(cancer_data['feature_names'])

['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']


In [43]:
print(cancer_data['data'])

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]


## Formatting Data

Next, let's create the dataframe to work with using pandas

In [44]:
df = pd.DataFrame(cancer_data['data'], columns=cancer_data['feature_names'])
df['target'] = cancer_data['target']    # Don't forget to put the target column as well
print(df.head())

   mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0             

## Training

The data is already well prepared and ready for training. To do so, we can use the model provided from scikit learn to train our data. We are also going to use a different solver for this, since it is going to take a lot of interation if we use the default solver.

In [45]:
X = df[cancer_data.feature_names].values
y = df['target'].values

model = LogisticRegression(solver='liblinear')
model.fit(X, y)
print(model.coef_, model.intercept_)

[[ 2.09074559  0.11629937 -0.06576082 -0.00296787 -0.1528625  -0.40087545
  -0.64119654 -0.33783008 -0.22302828 -0.02621972 -0.02607014  1.25233497
  -0.0043571  -0.09296504 -0.01676624  0.00361164 -0.04923056 -0.03999723
  -0.04219282  0.00608599  1.22945621 -0.33919543 -0.117277   -0.02427392
  -0.28329831 -1.13629672 -1.58028133 -0.65127888 -0.68868816 -0.11453176]] [0.39016927]


## Verify

Now let's verify the model by trying to predict the first 5 data

In [46]:
print(model.predict(X[:5]))

[0 0 0 1 0]


In [47]:
print(y[:5])

[0 0 0 0 0]


As we can see, our model succesfully predicts the 4/5 of the data. Next, we can try to find the accuracy score of our model.

In [48]:
print(model.score(X, y))

0.9595782073813708


This shows, that our model has 96% accuracy, which mean the predicted answer is 96% correct.